### PR TITLE
Added null escape to lexer

### PR DESCRIPTION
**Pull request by arthmoeros**
New PR (previous one -> https://github.com/vectordotdev/vrl/pull/219), had to make a new one because of conflicts.

Partially solves https://github.com/vectordotdev/vrl/issues/148

added supported null \0 escape character to new lexer

### CODE REVIEW

File: src/parser/lex.rs [hunk 1]
```diff
@@ -1308,6 +1308,7 @@ fn unescape_string_literal(mut s: &str) -> String {
                 b'n' => '\n',
                 b'r' => '\r',
                 b't' => '\t',
+                b'\0' => '\0',
```
Comment (fuchsnj) [inline src/parser/lex.rs:1311]:
```suggestion
                b'0' => '\0',
```
The character after the backslash is a literal `0` not null. This results in a panic (since it reaches the `unimplemented` below). The test you added was fine, but it only tests the lexer. The panic occurs in the parser (which runs after the lexer). I would suggest also adding a unit test to the `unescape_string_literal` function for the null character, and maybe consider adding it to our VRL test suite, in  a new file in `lib/tests/tests/expressions/literal` which will run it through the entirety of VRL.

I also want to point out you can use the VRL CLI (by running `cargo run` in `lib/cli`) when testing locally. This lets you easily test out new features like this.

Let me know if you need any help adding more tests for this. After this change, it appears to work for me in the CLI.

**Review by fuchsnj — CHANGES_REQUESTED** [review]

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -13,8 +13,9 @@ provide everything you need to get started.
    GitHub account (only applicable to outside contributors).
 3. [Create a new Git branch][urls.create_branch].
 4. Make your changes.
-5. Add a one line summary of changes to the "unreleased" section of `CHANGELOG.md`.
-6. [Submit the branch as a pull request][urls.submit_pr] to the repo. A team member should
+5. Add and/or update unit tests to cover your changes.
```
Comment (fuchsnj) [inline CONTRIBUTING.md:16]: Thanks for adding this. I suggest a minor edit here, since we have several different ways to test features. Not all of them are "unit" tests.
```suggestion
5. Add and/or update tests to cover your changes.
```

**Review by fuchsnj — APPROVED** [review]
This seems to be working. Thanks!
